atlasdb-commons/src/main/java/com/palantir/util/result/Result.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * (c) Copyright 2023 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.util.result;
+
+import com.palantir.logsafe.exceptions.SafeRuntimeException;
+
+public abstract class Result<T, E> {
```

Comment:
My bias would be two different approaches for this class: Have a concrete-final version of this class which exposes two static-constrcutors, or an interface which itself has two static constructors present. I.e.:

```
final class Result<T, E> {

    Optional<T> value;
    Optional<E> error;

    private Result(Optional<T> value, Optional<E> error) {
        this.value = value;
        this.error = error;
        Preconditions.checkArguments(value.isPresent() || error.isPrsent());
    }

    static Result<T, E> ok(T value);
    static Result<T, E> error(E error);

    public final boolean isError() {
        return error.isPresent();
    }

    public T unwrap() {
        return value.orElseThrow(() -> ...);
    }
    ...
}
```

or 
```
interface Result<T, E> {
    static Result<T, E> ok(T value);
    static Result<T, E> error(E error);
    
   boolean isOk();
   T unwrap();
   
   private static final class OkResult implements Result {...}
   private static final class ErrorResult implements Result {...}
}
```

Reply:
(2) is basically your current approach, but I would make `OkResult` implement `isOk` to avoid branching/instanceof. I would also hide the classes as they're an implementation detail, and have static constructors so users of the API can see what result types they can construct (especially since we want to keep this tightly scoped!)